File: NDICore/src/main/java/com/ndi/android/NDIlib.java
package com.ndi.android;

/**
 * Classe de interface Java para a biblioteca nativa NDI.
 */
public class NDIlib {

    static {
        try {
            System.loadLibrary("ndi"); // Carrega libndi.so
        } catch (UnsatisfiedLinkError e) {
            throw new RuntimeException("Não foi possível carregar a biblioteca NDI: " + e.getMessage());
        }
    }

    /**
     * Inicializa a biblioteca NDI.
     *
     * @return true se inicializou corretamente, false caso contrário
     */
    public static native boolean initialize();

    /**
     * Finaliza a biblioteca NDI.
     */
    public static native void destroy();

    /**
     * Cria um sender NDI.
     *
     * @param name Nome do sender
     * @return ponteiro nativo do sender como long
     */
    public static native long sendCreate(String name);

    /**
     * Destroi um sender NDI.
     *
     * @param sender ponteiro nativo do sender
     */
    public static native void sendDestroy(long sender);

    /**
     * Envia um frame de vídeo pelo sender NDI.
     *
     * @param sender ponteiro nativo do sender
     * @param frame  dados do frame em bytes
     * @param width  largura do frame
     * @param height altura do frame
     */
    public static native void sendVideo(long sender, byte[] frame, int width, int height);
}

File: app/src/main/java/com/zer0/zcam_ndi/NDIlib.kt
package com.zer0.zcam_ndi

object NDIlib {

    init {
        System.loadLibrary("ndi") // Lib gerada pelo CMake
    }

    // --- Funções JNI ---
    external fun initialize(): Boolean                        // Inicializa a lib NDI
    external fun sendCreate(name: String): Long              // Cria sender NDI
    external fun sendVideo(senderPtr: Long, width: Int, height: Int, data: ByteArray) // Envia vídeo
    external fun sendAudio(senderPtr: Long, channels: Int, sampleRate: Int, data: ShortArray) // Envia áudio
    external fun sendDestroy(senderPtr: Long)                // Libera sender
    external fun destroy()                                   // Finaliza biblioteca NDI

    // --- Wrappers Kotlin ---
    private var senderPtr: Long = 0

    fun createSender(name: String) {
        senderPtr = sendCreate(name)
        if (senderPtr == 0L) {
            throw RuntimeException("Falha ao criar sender NDI")
        }
    }

    fun sendFrame(data: ByteArray, width: Int, height: Int) {
        if (senderPtr == 0L) return
        sendVideo(senderPtr, width, height, data)
    }

    fun sendAudioFrame(channels: Int, sampleRate: Int, data: ShortArray) {
        if (senderPtr == 0L) return
        sendAudio(senderPtr, channels, sampleRate, data)
    }

    fun close() {
        if (senderPtr != 0L) {
            sendDestroy(senderPtr)
            senderPtr = 0
        }
    }
}
